xc/xc7/tests/mmcm/gen_random_cases.py

```diff
@@ -0,0 +1,257 @@
+
+        # Integer divider ranges from 2 to 64
+        else:
+            mult = random.randint(2, 64)
+
+        phase = random.uniform(0.0, +180.0)
+
+        # The phase shift requires to be quantized into 45 / mult steps.
+        quant = 45.0 / mult
+        phase = int((phase + quant / 2) / quant) * quant
+
+        return mult, phase
+
+    def gen_div_and_phase(frac_en):
+        """
+        For CLKOUTn
+        """
+
+        # Fractional multiplier ranges from 2.0 to 128.0. The step is 1/8.
+        if frac_en:
+
+            # Make sure to generate a fractional value
+            while True:
+                divide = random.uniform(2.0, 128.0)
+                divide = int(divide * 8.0) / 8.0
+
+                if divide != int(divide):
+                    break
+
+        # Integer divider ranges from 2 to 128
+        else:
+            divide = random.randint(1, 128)
+
+        phase = random.uniform(0.0, +180.0)
+
+        # The phase shift requires to be quantized into 45 / mult steps.
+        quant = 45.0 / divide
+        phase = int((phase + quant / 2) / quant) * quant
+
+        return divide, phase
+
+    params = {}
+
+    # Other settings
+    params["bandwidth"] = random.choice((
+        "LOW",
+        "HIGH",
+        "OPTIMIZED",
+    ))
+
+    # Input clock divider (1 to 106)
+    params["divclk_divide"] = random.randint(1, 106)
+
+    # Feedback clock
+    frac_en = random.random() > 0.5
+    mult, phase = gen_mult_and_phase(frac_en)
+
+    params["clkfbout_mult"] = mult
+    params["clkfbout_phase"] = phase
+
+    # Clock outputs
+    params["clkout"] = []
+    for i in range(7):
+
+        # Fractional divider is only available for CLKOUT0
+        frac_en = random.random() > 0.5
+        if i > 0:
+            frac_en = False
+
+        divide, phase = gen_div_and_phase(frac_en)
+
+        # Duty cycle must be 50% when fractional divider is enabled
+        duty = 0.5  # random.uniform(0.10, 0.90)
+        if frac_en:
+            duty = 0.5
+
+        params["clkout"].append(
+            ClkOut(
+                index=i,
+                enabled=random.random() > 0.20,
+                divide=divide,
+                duty=duty,
+                phase=phase
+            )
+        )
+
+    return params
+
+
+def validate_params(params):
+    """
+    Validates MMCM parameters agains legal VCO frequency range. Makes Vivado
+    don't complain.
+    """
+
+    # VCO operating ranges [MHz] (for xc7a35tcsg324-1 speed grade -1)
+    vco_range = (600.0, 1200.0)
+
+    mul = params["clkfbout_mult"]
+    div = params["divclk_divide"]
+
+    # It is impossible to meet VCO freq. constraints both for 100MHz and 50Mhz
+    # input. Hence we check only for 100.
+    for f_clkin in (100.0, ):
+
+        f_vco = f_clkin * mul / div
+
+        if f_vco < vco_range[0]:
+            return False
+        if f_vco > vco_range[1]:
+            return False
+
+    return True
+
+
+def make_integers(params):
```

Comment:
This function shouldn't be needed here any more?

Reply:
I've removed all code that converts MMCM parameters.